shared/typetracking/codeql/typetracking/TypeTracking.qll

```diff
@@ -130,499 +301,510 @@ private module Cached {
-   */
   pragma[inline]
-  predicate smallstep(Node nodeFrom, TypeTrackingNode nodeTo, StepSummary summary) {
-    smallstepNoCall(nodeFrom, nodeTo, summary)
-    or
-    smallstepCall(nodeFrom, nodeTo, summary)
+  private predicate smallStepSplit(Node nodeFrom, Node nodeTo, StepSummary summary) {
+    smallStepCall(nodeFrom, nodeTo, summary) or smallStepNoCall(nodeFrom, nodeTo, summary)
   }
 
-  deprecated predicate localSourceStoreStep = flowsToStoreStep/3;
-}
-
-/**
- * A summary of the steps needed to track a value to a given dataflow node.
- *
- * This can be used to track objects that implement a certain API in order to
- * recognize calls to that API. Note that type-tracking does not by itself provide a
- * source/sink relation, that is, it may determine that a node has a given type,
- * but it won't determine where that type came from.
- *
- * It is recommended that all uses of this type are written in the following form,
- * for tracking some type `myType`:
- * ```ql
- * DataFlow::TypeTrackingNode myType(DataFlow::TypeTracker t) {
- *   t.start() and
- *   result = < source of myType >
- *   or
- *   exists (DataFlow::TypeTracker t2 |
- *     result = myType(t2).track(t2, t)
- *   )
- * }
- *
- * DataFlow::Node myType() { myType(DataFlow::TypeTracker::end()).flowsTo(result) }
- * ```
- *
- * Instead of `result = myType(t2).track(t2, t)`, you can also use the equivalent
- * `t = t2.step(myType(t2), result)`. If you additionally want to track individual
- * intra-procedural steps, use `t = t2.smallstep(myCallback(t2), result)`.
- */
-class TypeTracker extends TTypeTracker {
-  Boolean hasCall;
-  OptionalTypeTrackerContent content;
-
-  TypeTracker() { this = MkTypeTracker(hasCall, content) }
-
-  /** Gets the summary resulting from appending `step` to this type-tracking summary. */
-  TypeTracker append(StepSummary step) { result = append(this, step) }
-
-  /** Gets a textual representation of this summary. */
-  string toString() {
-    exists(string withCall, string withContent |
-      (if hasCall = true then withCall = "with" else withCall = "without") and
-      (
-        if content != noContent()
-        then withContent = " with content " + content
-        else withContent = ""
-      ) and
-      result = "type tracker " + withCall + " call steps" + withContent
-    )
+  pragma[inline]
+  private predicate stepSplit(Node nodeFrom, Node nodeTo, StepSummary summary) {
+    stepNoCall(nodeFrom, nodeTo, summary) or stepCall(nodeFrom, nodeTo, summary)
   }
 
-  /**
-   * Holds if this is the starting point of type tracking.
-   */
-  predicate start() { hasCall = false and content = noContent() }
-
-  /**
-   * Holds if this is the starting point of type tracking, and the value starts in the content named `contentName`.
-   * The type tracking only ends after the content has been loaded.
-   */
-  predicate startInContent(TypeTrackerContent contentName) {
-    hasCall = false and content = contentName
+  pragma[nomagic]
```

Comment:
Shouldn't this be a `pragma[inline]`?

Reply:
That's currently left for a future PR. I believe the answer will end up being no, and these internal predicates can maybe be simplified a bit, as a call-graph-recursive use-case will need a slightly different `step` member predicate.

Reply:
👍